client/services/code_preprocessors/PythonCodePreprocessorService.js

```diff
@@ -152,117 +160,183 @@ tie.factory('PythonCodePreprocessorService', [
       };
 
       return testCode;
     };
 
     var _generateBuggyOutputTestCode = function(
-        buggyOutputTests, inputFunctionName, outputFunctionName) {
+        allTasksBuggyOutputTests,
+        allTasksInputFunctionNames, allTasksOutputFunctionNames) {
       // NOTE: This must be run after the correctness tests, since it assumes
       // that the test inputs and correctness test results already exist.
       // TODO(sll): Cache the results of running the buggy code, so that they
       // don't have to be recomputed for every run.
-      var testInputCode = (
-        inputFunctionName ?
-        inputFunctionName + '(test_input)' :
-        'test_input'
-      );
-      var testRunCode = 'System.runTest(func, ' + testInputCode + ')';
-      var testOutputCode = (
-        outputFunctionName ?
-        outputFunctionName + '(' + testRunCode + ')' : testRunCode);
-
-      var fullTestCode = [
-        'def matches_buggy_function(func):',
-        '    buggy_results = []',
-        '    for test_input in test_inputs:',
-        '        buggy_results.append(' + testOutputCode + ')',
-        '    return buggy_results == ' + VARNAME_CORRECTNESS_TEST_RESULTS,
-        '',
-        VARNAME_BUGGY_OUTPUT_TEST_RESULTS + ' = []',
-        ''
-      ].join('\n');
-      buggyOutputTests.forEach(function(buggyOutputTest) {
-        var qualifiedBuggyFunctionName = (
-          buggyOutputTest.getBuggyFunctionName());
-
-        fullTestCode += (
-          VARNAME_BUGGY_OUTPUT_TEST_RESULTS +
-          '.append(matches_buggy_function(' +
-          qualifiedBuggyFunctionName + '))\n');
-      });
+      var matchesBuggyFunctionSet = new Set();
+
+      for (var i = 0; i < allTasksInputFunctionNames.length; i++) {
+        var inputFunctionName = allTasksInputFunctionNames[i];
+        var outputFunctionName = allTasksOutputFunctionNames[i];
+
+        var matchesBuggyFunctionName = 'matches_buggy_function' +
+            (inputFunctionName == null? '' : '_' + inputFunctionName) +
+            (outputFunctionName == null? '' : '__' + outputFunctionName) ;
+
+        if (!matchesBuggyFunctionSet.has(matchesBuggyFunctionName)) {
+          matchesBuggyFunctionSet.add(matchesBuggyFunctionName);
+          var testInputCode = (
+            inputFunctionName ?
+            inputFunctionName + '(test_input)' :
+            'test_input'
+          );
+          var testRunCode = 'System.runTest(func, ' + testInputCode + ')';
+          var testOutputCode = (
+            outputFunctionName ?
+            outputFunctionName + '(' + testRunCode + ')' : testRunCode);
+
+          var fullTestCode = [
+            'def ' + matchesBuggyFunctionName + '(func):',
+            '    buggy_results = []',
+            '    for task_tests in all_tasks_test_inputs:',
+            '        task_results = []',
+            '        for test_input in task_tests:',
+            '            task_results' +
+            '.append(' + testOutputCode + ')',
```

Comment:
Go ahead and put this on the previous line, even if it exceeds the line limit. It is much easier to read that way.